keccak256/src/gates/absorb.rs

```diff
@@ -87,13 +80,53 @@ impl<F: FieldExt> AbsorbConfig<F> {
         }
     }
 
+    pub fn assign_next_inp_and_flag(
+        &self,
+        region: &mut Region<F>,
+        offset: usize,
+        flag: (Cell, F),
+        next_input: [F; NEXT_INPUTS_LANES],
+    ) -> Result<(Cell, F), Error> {
+        // Generate next_input in base-9.
+        let mut next_mixing = state_to_biguint::<F, NEXT_INPUTS_LANES>(next_input);
+        for (x, y) in (0..5).cartesian_product(0..5) {
+            if x >= 3 && y >= 1 {
+                break;
+            }
+            next_mixing[(x, y)] = convert_b2_to_b9(next_mixing[(x, y)].clone().try_into().unwrap())
+        }
+        let next_input = state_bigint_to_field::<F, NEXT_INPUTS_LANES>(next_mixing);
+
+        // Assign next_mixing at offset = 1
```

Comment:
The offset increment is now done outside this helper.

```suggestion
        // Assign next_mixing.
```